Bugfix/166 invest logging stdout

davemfish opened the pull request:
This fixes natcap/invest-workbench#166

We're splitting the display of invest logfiles into two distinct modes:
1. During an active invest run, the invest log is displayed by listening to the invest process stdout & stderr
2. To display the log of a pre-existing invest run loaded from the list of "Recent Runs", read the logfile from disk

We no longer need to worry about watching for logfiles to be created in active workspaces, or with watching the file itself for updates during an active run.

There's also some new tracking of the state of an invest subprocess so we can display the appropriate components. For example, the 'Cancel' button should only be available after we are certain the invest process is running and has a `pid`. But the 'Run' button should be disabled even earlier, as soon as it is clicked, and not enabled again until invest exits.

Also in this PR, a bit of debugging for the puppeteer test that should make it more resilient to things that seem to take varying amounts of time, like the electron startup, and invest validation.

davemfish committed:
read invest logging from stdout instead of from the logfile, for active invest runs. #166.

davemfish committed:
removed tests for the findLogfile function that we no longer need. #166.

davemfish committed:
simplified the invest subprocess tests because they no longer need a real workspace with logfile. #166.

davemfish committed:
added test to cover invest logging switching from a pre-existing file to a new invest stdout stream. #166.

davemfish committed:
use core node modules to read the logfile in the main process. #166.

davemfish committed:
added error handling to the logfile reader. #166.

davemfish committed:
removed the Tail package and related functions from LogTab. #166.

davemfish committed:
fixed bug in fetch-invest script where multiple versions of python packages could accumulate. #166.

davemfish committed:
added state to specifically control state of execute button separate from invest process status, which controls state of cancel button, etc. #166.

davemfish committed:
debugged error re-loading a logfile. #166.

davemfish committed:
switched away from a pptr-testing-library query that wasn't working. #166.

davemfish committed:
cleanup in pptr script to make it more resilient to pre-existing local AppData. #166.

davemfish committed:
trying to avoid a common npm install error in GHA. #166.

davemfish committed:
Revert "trying to avoid a common npm install error in GHA. #166."

This reverts commit 78bac2a32ec06552966636f45a566d918aa0cb94.

davemfish committed:
updated to jest v27. #164

davemfish committed:
Revert "updated to jest v27. #164"

This reverts commit 6543db759e71d1870a7cbcc9551a141f410ed44c.

davemfish committed:
resolved conflicts in app.test.js. #166.

davemfish committed:
debugged pptr script - use a selector that will wait. #166.

davemfish committed:
debugged the main process teardown functions that were not firing on macos and maybe not on windows either. #166.

davemfish committed:
improved pptr connection to electron window by waiting for a signal from electron that it's ready to connect. #166.

davemfish committed:
extending the timeout to hear from electron that it's ready for pptr. #166.

davemfish committed:
Merge remote-tracking branch 'upstream/main' into bugfix/166-invest-logging-stdout

davemfish committed:
listen for and log page crashes with puppeteer. #166.

davemfish committed:
udpated a comment. #166.

davemfish committed:
updated to latest puppeteer-core, even though it's not guaranteed to work with electronv13's chromium version. #166.

davemfish committed:
removed a console.log. #166.

davemfish committed:
Revert "updated to latest puppeteer-core, even though it's not guaranteed to work with electronv13's chromium version. #166."

This reverts commit 87b2af65696667b340e2dd9549daa510f640f714.

davemfish committed:
cleaning up some comments and console logs. #166.

davemfish committed:
added listener for a renderer process crash, for better debugging. #166.

davemfish committed:
enable electron crashReporter and place logs in the same folders as puppeteer screenshots. #166.

davemfish committed:
need a mock for electron.crashReporter. #166.

davemfish committed:
need a mock for electron.crashReporter. #166.

davemfish committed:
Merge branch 'bugfix/166-invest-logging-stdout' of github.com:davemfish/invest-workbench into bugfix/166-invest-logging-stdout

davemfish committed:
some debug logging for the crashReporter. #166.

davemfish committed:
fixing electron mocks. #166.

davemfish committed:
need to set the crashReporter path before starting the reporter. #166.

davemfish committed:
integrate crashReporter with free trial of backtrace.io. #166.

davemfish committed:
migrating to yarn v2, but keeping node_modules mode. #166

davemfish committed:
fixed typo in yarnrc. #166

davemfish committed:
migrated yarn.lock to v2 format. #166.

davemfish committed:
forgot to checkin the .yarn folder. #166.

davemfish committed:
reverting all the yarn v2 changes. #166.

davemfish committed:
attempting to workaround occasional yarn install timeouts. #166.

davemfish committed:
more debugging with event listeners for main window. #166.

davemfish committed:
added extra timeout where we need to wait for invest validation to resolve. #166.

davemfish committed:
attempting to get a new error message by removing afterAll code. #166.

davemfish committed:
making sure the test-level timeout is longer than the sum of all query timeouts. #166.

davemfish committed:
removed a setTimeout pause in test that doesn't seem necessary anymore. #166.

davemfish committed:
some linting. #166.

davemfish committed:
fixing some missing props in tests. #166.

davemfish committed:
removed the electron crashReporter stuff since we were not actually experiencing any crashes. #166.

davemfish committed:
removed some listeners that were used for debugging. #166.

davemfish committed:
added to a test to confirm the cancel button is only available after signal is sent from main to renderer. #166.

davemfish committed:
removed the crashReporter mock we're no longer using. #166.

davemfish committed:
fixed logic in control flow during app before-quit. #166.

phargogh commented on src/main/main.js:
```diff
@@ -106,21 +108,22 @@ export const createWindow = async () => {
     if (ELECTRON_DEV_MODE) {
       mainWindow.webContents.openDevTools();
     }
+    // We use this stdout as a signal in a puppeteer test
+    process.stdout.write('main window loaded');
```
Neat!  Very helpful comment, too.

phargogh commented on src/main/setupInvestHandlers.js:
```diff
@@ -127,31 +90,38 @@ export function setupInvestRunHandlers(investExe) {
+      if (!investLogfile) {
+        if (`${data}`.match('Writing log messages to')) {
+          investLogfile = `${data}`.split(' ').pop().trim();
+          runningJobs[args.workspace_dir] = investRun.pid;
```
Out of curiosity, is it possible for a user to start 2 model runs (could be different models) in the same workspace with different PIDs?  If it is, wouldn't we lose track of the PID of the first process?

davemfish replied:
Oh thanks for catching this! I switched the key to the `jobID` that is unique to each `InvestTab` component. Since we know we can only have one running job per open tab, that should fix this.

phargogh replied:
Ah right!  That makes a lot of sense.  Thanks!

phargogh commented on src/main/setupInvestHandlers.js:
```diff
@@ -166,3 +136,25 @@ export function setupInvestRunHandlers(investExe) {
+
+    const rl = readline.createInterface({
+      input: fileStream,
+      crlfDelay: Infinity,
```
Well this is an interesting option!  I wonder what their use case was for expecting there to be an observable delay between `\r` and `\n`.

davemfish replied:
I guess it must have happened at some point! Based on the docs `Infinity` seems like a reasonable cross-platform choice without any downside in our case.

phargogh commented on src/main/setupInvestHandlers.js:
```diff
@@ -166,3 +136,25 @@ export function setupInvestRunHandlers(investExe) {
+      );
+    });
+
+    const rl = readline.createInterface({
```
I'm sure this is just my relative node inexperience coming out, but I'm not sure I understand the benefit of having a readline interface atop `fileStream` ... is it just that we're trying to iterate one line at a time rather that whatever chunks `fileStream` would normally read?  Is there a limit to the size of the data that can be sent across the `invest-stdout-*` channel?  Something else?  Sorry for my confusion here!

davemfish replied:
@phargogh That's a great question that I never thought too deeply about. The goal is to read one line at a time because that's how the `LogTab` `markupLine` is designed to work. If you don't mind, could you take a look in `LogTab` and give that design a bit of critical review? I wouldn't want to design this whole `fileStream + readline` stuff around something else that was a flawed idea to begin with.

At some point we'll probably move that `markupLine` feature out of the renderer and into this main process because it's still using one node module (`santizeHtml`) and we ought to eventually remove node integration from the renderer completely. That could mean just substituting in another `ipc` call.

phargogh replied:
Sure!  So, I'm not certain I can speak to whether `markupLine` belongs in the main process or in the renderer at the moment, but your reasoning makes sense to me about that.

The main thing I'm noticing about this design centers around the line-by-line nature of it.  That's not _inherently_ a problem, but the fact that we start from the very beginning with reading from stdout line-by-line is probably less efficient than it could be.  The [Node docs for `createReadStream`](https://nodejs.org/api/fs.html#fs_fs_createreadstream_path_options) suggest that it'll by default read in 16Kb at a time, which is usually a whole lot more than a single line!  The point of deliberately passing around larger chunks is that it tends to reduce iteration overhead, and since each chunk here will be passing through the IPC (where presumably there's some overhead per call), I'd wonder about all of that overhead just generally slowing things down.

So if I were to make an architectural observation, it would be that we might be able to reduce overhead by reading data in larger chunks (the default should be fine) and then only split by newlines where it's needed so that we minimize that communication overhead.  It might not make a difference with just running 1 model, but if we're trying to monitor a bunch of streams all at the same time, I bet that would become noticeable.

And maybe this wouldn't matter at all if we move `markupLine` out of the renderer and we'll need more IPC for that move?  I'm not sure.

And I could of course be totally off base with this ... [my timings](https://gist.github.com/phargogh/00aea868fddcab3def00840da5482dcb) were around simply reading from a file stream, which is fairly simple compared with inter-process communication.  Anyways, I'd be curious what you think about this.

phargogh replied:
We talked about this on the office hours huddle today and decided that there isn't enough of a slowdown to warrant a complete refactor of the line-by-line handling of markup, and we can always try to optimize it later on when we do have a clear slowdown.  @davemfish was also going to try using python's `PYTHONUNBUFFERED=1` environment variable to see how well that works and whether that might be useful and/or faster within the workbench context of reading from stdout.

davemfish replied:
@phargogh here's an update. Adding the `PYTHONUNBUFFERED` env variable seems to work. There's no discernable difference with or without it for the invest runs I've tried, but this still seems like a good use-case for it.

However, reading line-by-line from a file on disk takes too long for even a modest-sized logfile of 1660 lines. It takes a few seconds to load. 

After more thinking, it might be reasonable to only do the markup during active invest runs as the data comes through stdout. Then for loading "Recent Run" logfiles, we just load it all as plaintext with no markup. In the future, if we want to add the markup feature to "Recent Run" logs, we might consider caching the marked-up logfile that we create in-memory during the invest run, and load *that* back later instead of the plaintext logfile. What do you think?

davemfish replied:
Ditching line-by-line would also solve another minor problem: reading logfiles line-by-line is not exactly the same as logger message-by-message. As some messages contain newlines. So there were occasional cases where the pattern-matcher only marked-up the first line of a multi-line logger message. And our regex got more complicated in order to deal with that, especially for Tracebacks.

phargogh replied:
> And our regex got more complicated in order to deal with that, especially for Tracebacks.

Yeah, I was wondering about this case as well.  Remembering back to that PyBay talk we had both attended, would be be worth considering structured logging?

> However, reading line-by-line from a file on disk takes too long for even a modest-sized logfile of 1660 lines. It takes a few seconds to load.
>
> After more thinking, it might be reasonable to only do the markup during active invest runs as the data comes through stdout. 

Oh my, that's a notable delay on a not-terribly long file!  I like your solution of marking up the file, caching it and presenting that to the user.  That feels like a good balance of orthogonality, keeping the UI features in the UI layer and still minimizing the need to recompute the HTML markup on file load.

phargogh reviewed (commented):
Thanks @davemfish !  I've looked through and had a couple questions for my own information that are very likely just due to my relative inexperience with node, but I'm curious what you think.

davemfish committed:
switched to an identifier unique to each InvestTab component for mapping active invest runs to the subprocess pid. #166.

davemfish committed:
moved the markup of invest log messages to the main process because it relies on a node package. #166.

davemfish committed:
when loading an invest log from file, don't read line-by-line or markup with html. #166.

davemfish committed:
skipping tests for markup of invest logs when reading log from file. #166.

davemfish committed:
added unit tests for the logfile markup function; skipping tests that assert markup on pre-existing file. #169 and #166.

davemfish commented:
Okay @phargogh I think this is ready for review again. I made #169 based on the outcome of our discussion. 

As for structured logging in invest vs. parsing plaintext with regex...I don't think we need to change invest logging for the benefit of the Workbench right now. Generally I think invest logs should stay nice & readable as text files and adding more structure to them might make them less readable.

davemfish commented on src/main/setupInvestHandlers.js:
```diff
@@ -112,50 +102,66 @@ export function setupInvestRunHandlers(investExe) {
+    const stdErrCallback = (data) => {
+      // The python Traceback for invest comes through stdout & stderr,
+      // So no need to merge those streams for the benefit of displaying
+      // a complete log.
```
This was an unexpected realization. It's convenient, but do we know exactly why this happens. And also there's at least one case where the Traceback is *not* going through stdout. https://github.com/natcap/invest/issues/651

phargogh replied:
Based on what I found while working on https://github.com/natcap/invest/pull/655, the intent of the way InVEST's logging is structured in a `--headless` run is that logging will be sent to 2 handlers: `StreamHandler(sys.stdout)` and a `FileHandler`.  So a `LOGGER.exception` call should appear in both, and an uncaught exception (the CLI doesn't handle an exception in `execute`) will be printed to `stderr` as you'd expect for a python program.  It does seem like there's some undefined behavior around whether an exception message would be written to `stdout` via logging before the application terminates and the exception is written to `stderr`.

The changes in https://github.com/natcap/invest/pull/655 _should_ make exception handling more consistent and should clarify to python that we want to log the exception in the CLI if `execute` raises one.

phargogh reviewed (changes requested):
Makes sense to me!  I had a couple of very minor comments and a few questions, but nothing major I hope.

davemfish committed:
fixed bug in scoping of a variable. #166.

davemfish committed:
fixed bug where duplicate ipcRenderer listeners were registered everytime execute button was clicked. #166.

davemfish committed:
guarding against a pattern that is not a RegExp in the markup function. #166.

davemfish commented on src/main/setupInvestHandlers.js:
```diff
@@ -20,52 +20,42 @@ const LOGLEVELMAP = {
+const LOG_ERROR_REGEX = /(Traceback)|(([A-Z]{1}[a-z]*){1,}Error)|(ERROR)/;
+export const LOG_PATTERNS = {
+  'invest-log-error': LOG_ERROR_REGEX,
+  'invest-log-primary': /a^/, // default is regex that will never match
```
That's a good point. I changed this default value and the function to skip the pattern matching if the pattern is not a regex.

davemfish commented on src/main/setupInvestHandlers.js:
```diff
@@ -112,50 +102,66 @@ export function setupInvestRunHandlers(investExe) {
+    let investLogfile;
+    const env = {
+      PATH: path.dirname(investExe),
+      PYTHONUNBUFFERED: 'TRUE',
```
I actually removed this env var since discovering that `logging` flushes with each emit. Since our goal is to get each logger message one at a time over stdout, I think that's all we need! Do you concur? https://github.com/python/cpython/blob/main/Lib/logging/__init__.py#L1092

phargogh replied:
Yep, I concur!  That should do the trick for us here.

davemfish commented on src/main/setupInvestHandlers.js:
```diff
@@ -112,50 +102,66 @@ export function setupInvestRunHandlers(investExe) {
+        markupMessage(`${data}`, LOG_PATTERNS)
+      );
+    };
+    LOG_PATTERNS['invest-log-primary'] = new RegExp(pyModuleName);
```
Good catch, these would be in conflict. I fixed that bug.

davemfish commented on src/renderer/components/InvestTab/index.jsx:
```diff
@@ -110,18 +113,21 @@ export default class InvestTab extends React.Component {
     });
     ipcRenderer.on(`invest-stderr-${jobID}`, (event, data) => {
+      // It's convenient to have stderr in it's own object to display
+      // it an Alert.
```
I moved this comment to somewhere where I think it's needed more.

davemfish committed:
added to a test to cover filtering the PyInstaller stderr message. #166.

davemfish committed:
removed some unused imports in a test. #166.

davemfish commented:
This should also fix #131 . Some of the trouble resolving that has been a bug I just discovered and fixed here:

We were registering `ipcRenderer.on` listeners every time the Run button is clicked. Which means it was easy to get duplicate listeners for the data being sent from `investRun.stderr.on('data', stdErrCallback)`, for example. So I think that was polluting the stderr strings that we trying to pattern-match on to filter out that `Failed to execute script ....` message from pyinstaller. I added to a test to cover filtering out that message.

phargogh reviewed (approved):
Thanks Dave!
